Studio: Add prompt limit UI

kozer opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Fixes https://github.com/Automattic/dotcom-forge/issues/7409

## Proposed Changes

This diff, adds a prompt limit UI into user settings

## Testing Instructions
1. Start studio `nvim use && STUDIO_AI=true npm start`
3. Ensure you are logged in
4. Navigate to settings and ensure that you see the following UI screen
![2024-05-31T10:25:46,590050296+03:00](https://github.com/Automattic/studio/assets/497103/f29481ef-96f0-4b5d-8555-a50a71a7be7d)
5. Start studio without STUDIO_AI feature enabled. Ensure that the new UI is not shown.

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

kozer committed:
Studio: Add prompt limit UI

kozer committed:
Update: Add defaults, and reject on error

kozer committed:
Update: Minor refactor on how we handle errors

kozer committed:
Update: Move to context in order to share state across components

kozer committed:
Update: Fix lint errors

derekblank reviewed (approved):
LGTM. I was able to observe prompt limit changes within the Settings screen.

<img width="405" alt="Screenshot 2024-06-04 at 12 09 40 pm" src="https://github.com/Automattic/studio/assets/643285/676e1369-d4d2-4954-96a0-1a787188ac19">

kozer committed:
Merge branch 'trunk' into add/setting-prompt-limit-ui

kozer committed:
Update: Not show the prompt limit if ai feature is not enabled

kozer committed:
Update: fix lint issues

sejas commented on src/hooks/use-prompt-usage.tsx:
```diff
@@ -0,0 +1,90 @@
+			Sentry.captureException( error );
+			console.error( error );
+		}
+	}, [ client ] );
```
Should we add `assistantEnabled` to the dependency array?

I'm getting this lint warning:


<img width="1241" alt="Screenshot 2024-06-04 at 11 19 57" src="https://github.com/Automattic/studio/assets/779993/8d7c33e4-46c6-4152-9919-cea7c8d5c346">



```suggestion
	}, [ assistantEnabled, client ] );
```

sejas commented on src/components/content-tab-assistant.tsx:
```diff
@@ -57,6 +59,7 @@ export function ContentTabAssistant( { selectedSite }: ContentTabAssistantProps
 				] );
 				if ( message ) {
 					addMessage( message, 'assistant' );
+					fetchPromptUsage();
```
I noticed that we are making two requests, one to grab the message and this `fetchPrmoptUsage` to fetch the Headers and update the current usage.
An optimization would be getting the Headers from `fetchAssistant` and then just updating he usage here without making an extra call. It's not a big deal if we make two requests, but something to consider.

sejas reviewed (commented):
@kozer , the happy path works great.
We need to fix the unit tests.
I'm seeing this error `Error: Uncaught [TypeError: Cannot read properties of undefined (reading 'assistantEnabled')]` in `src/components/tests/user-settings.test.tsx`

I suggested using the headers from `fetchAssistant` to update the current usage, but its not a strong opinion.

One last thing I noticed is that when starting the app offline I see the count of my demo sites but the count of my prompts goes to zero. Would be possible to keep that number for consistency, or refetch it when I'm back online?
If yo prefer to handle that edge case in a separate PR I can create a new issue.

Happy Path:


https://github.com/Automattic/studio/assets/779993/bfdb1cd9-227d-4e60-961e-f3ceeab82df8

Steps to reproduce Offline edge case:

1. Disconnect your wifi/internet
2. Run the app
3. Open the settings modal by clicking on your user icon
4. Observe your prompts indicate 0 when it should be some number.

katinthehatsite commented on src/components/user-settings.tsx:
```diff
@@ -72,13 +74,11 @@ const SnapshotInfo = ( {
-			<h2 className="a8c-subtitle-small">{ __( 'Usage' ) }</h2>
-			<div className="flex gap-4 flex-row items-center w-full">
+		<div className="flex gap-3 flex-col">
+			<h2 className="a8c-label-semibold">{ __( 'Demo Sites' ) }</h2>
```
```suggestion
			<h2 className="a8c-label-semibold">{ __( 'Demo sites' ) }</h2>
```

It seems to be in the lower case in the screenshot on https://github.com/Automattic/dotcom-forge/issues/7409

kozer replied:
Updated! Thanks!

katinthehatsite commented on src/components/user-settings.tsx:
```diff
@@ -140,6 +140,33 @@ const SnapshotInfo = ( {
+	}
+	return (
+		<div className="flex gap-3 flex-col">
+			<h2 className="a8c-label-semibold">{ __( 'AI Assistant' ) }</h2>
```
This is also lowercase in the screenshot in https://github.com/Automattic/dotcom-forge/issues/7409 
But now I am wondering if it should be something like Studio WordPress Assistant? I am not sure because in the chat placeholder, this is what it looks like:
<img width="679" alt="Screenshot 2024-06-04 at 1 44 35 PM" src="https://github.com/Automattic/studio/assets/25575134/80106669-55dc-438e-9b68-73fa90c38803">

derekblank replied:
Just noting that in the [AIInput PR](https://github.com/Automattic/studio/pull/202/files#diff-37686ba47ce954cd13aa8a3f244d4a912cb652167b4ead0781b39caf6620813aR87), this text has been updated to "What would you like to learn?" -- the "Ask Studio WordPress Assistant" text was _truly_ a placeholder. 🙃 

The question of capitalizing assistant or Assistant is valid, though. I have been defaulting to capital-A Assistant per the name of the feature in the tab.

kozer replied:
I updated the texts as requested. I left the "AI assistant" as is for now! Thank you Kat!

kozer committed:
Merge branch 'trunk' into add/setting-prompt-limit-ui

kozer commented:
> If yo prefer to handle that edge case in a separate PR I can create a new issue.

Thank you for your comments!
Yes, indeed I did a last minute change to not show the bar when the feature is not enabled and the tests where broken. I'm working on that.

>If yo prefer to handle that edge case in a separate PR I can create a new issue.

Yes, let's create another issue to handle that. Thanks!

kozer committed:
Update: Use post to fetch headers, instead of doing and extra HEAD request

kozer committed:
Update: Fix tests

kozer committed:
Update: Lowercase settings headers

kozer commented:
@sejas , I updated the PR to use assistant API to fetch the headers, fixed tests, and did a minor refactor! Thanks for your suggestions!

cc: @katinthehatsite

sejas commented on src/components/content-tab-assistant.tsx:
```diff
@@ -7,6 +7,7 @@ import { useAssistant } from '../hooks/use-assistant';
 import { useAssistantApi } from '../hooks/use-assistant-api';
 import { useAuth } from '../hooks/use-auth';
 import { useOffline } from '../hooks/use-offline';
+import { usePromptUsage } from '../hooks/use-prompt-usage';
```
Let's remove this unused import.

sejas commented on src/components/user-settings.tsx:
```diff
@@ -72,13 +74,11 @@ const SnapshotInfo = ( {
 	const offlineMessage = __( 'Deleting demo sites requires an internet connection.' );
 	return (
-		<div className="flex gap-5 flex-col">
-			<h2 className="a8c-subtitle-small">{ __( 'Usage' ) }</h2>
```
~We shouldn't change the old behaviour until we release the feature flag.~
We are changing the old behaviour but that's ok in this case.

sejas commented on src/hooks/use-prompt-usage.tsx:
```diff
@@ -0,0 +1,103 @@
+			return;
+		}
+		try {
+			await new Promise( ( resolve, reject ) => {
+				client.req.get(
+					{
+						method: 'HEAD',
+						path: '/studio-app/ai-assistant/chat',
+						apiNamespace: 'wpcom/v2',
+					},
+					( error: Error, _data: unknown, headers: Record< string, string > ) => {
+						if ( error ) {
+							return reject( error );
+						}
+						if ( ! headers ) {
+							reject( new Error( 'No headers in response' ) );
+							return;
+						}
+						updatePromptUsage( headers );
+						resolve( headers );
+					}
+				);
+			} );
```
Do we need to create this promise and resolve the headers? Maybe we can await the `client.req.get` directly.

```suggestion
			await client.req.get(
				{
					method: 'HEAD',
					path: '/studio-app/ai-assistant/chat',
					apiNamespace: 'wpcom/v2',
				},
				( error: Error, _data: unknown, headers: Record< string, string > ) => {
					if ( error ) {
						return;
					}
					if ( ! headers ) {
						return;
					}
					updatePromptUsage( headers );
				}
			);
```

katinthehatsite commented:
Hi @kozer - thanks for making updates! The changes look good on my end.

I tested for this scenario:

1.Disconnect your wifi/internet
2. Run the app
3. Open the settings modal by clicking on your user icon
4. Observe your prompts indicate 0 when it should be some number.

And after updates, I can see that the number goes back up once I start the app back after being offline which is great. Would it be possible to keep it there even when offline? (At the moment, it goes to 0 when offline)

Otherwise, everything looks 👍

kozer commented:
> Hi @kozer - thanks for making updates! The changes look good on my end.
> 
> 
> 
> I tested for this scenario:
> 
> 
> 
> 1.Disconnect your wifi/internet
> 
> 2. Run the app
> 
> 3. Open the settings modal by clicking on your user icon
> 
> 4. Observe your prompts indicate 0 when it should be some number.
> 
> 
> 
> And after updates, I can see that the number goes back up once I start the app back after being offline which is great. Would it be possible to keep it there even when offline? (At the moment, it goes to 0 when offline)
> 
> 
> 
> Otherwise, everything looks 👍 
> 
> 
> 
> 

As discussed above, in Antonio's comment, I'll tackle this in a different pr ( the offline stuff ). 
Is it ok?

sejas commented on src/hooks/use-assistant-api.ts:
```diff
@@ -16,20 +18,40 @@ export function useAssistantApi() {
 			}
 			const message = response?.choices?.[ 0 ]?.message?.content;
-
+			updatePromptUsage( headers );
```
Thanks for optimizing this ! 🙏

katinthehatsite commented:
Sounds good @kozer ! I approved the PR ✅ Nice work!!

sejas reviewed (commented):
The PR looks great. I tested it I was able to see my usage updated. I noticed that the limit was restored to 0 from yesterday.

My only concern is that we are changing the layout for current users even if we didn't released our feature flag yet.

@wojtekn , what do you think about submitting this change that affects current users?
I think this specific change is harmless, but I understand if we prefer to keep the old demo site limit bar intact.

| Before |
|--------|
|  <img width="1012" alt="settings-usage-original" src="https://github.com/Automattic/studio/assets/779993/fd3716bc-7870-43dc-89d7-6321e557e15b"> | 

| After |
|--------|
|  <img width="1012" alt="settings-usage-changed" src="https://github.com/Automattic/studio/assets/779993/f391cffa-8435-4210-8013-057d1002d21d"> |

wojtekn commented:
> My only concern is that we are changing the layout for current users even if we didn't released our feature flag yet.
> @wojtekn , what do you think about submitting this change that affects current users?
I think this specific change is harmless, but I understand if we prefer to keep the old demo site limit bar intact.

It's okay. We are slightly changing the layout to prepare it for incoming changes, but we are not enabling AI assistant UI in any form.

sejas reviewed (approved):
🚀 
@kozer , feel free to merge after adressing:
- https://github.com/Automattic/studio/pull/197#discussion_r1627942600
- https://github.com/Automattic/studio/pull/197#discussion_r1627956210

sejas committed:
Merge branch 'trunk' of github.com:Automattic/studio into add/setting-prompt-limit-ui

sejas committed:
Remove unused import

sejas committed:
Simplify fetchPromptUsage removing unnecessary Promise

sejas reviewed (approved):
I applied these changes and tested it again. It works as expected. Thank you!
As we discussed we can update the headers name later.

Changes:
- https://github.com/Automattic/studio/pull/197/commits/e77d422b3782b3ef25e41ac30e88141d85e376fb
- https://github.com/Automattic/studio/pull/197/commits/9c79ffa1563bb289cdf68c1ad85408a737417cd7

https://github.com/Automattic/studio/assets/779993/df1b39e9-e24b-499b-a243-48c9e1314eae
